pkg/addonmanager/controllers/agentdeploy/controller.go

```diff
@@ -237,44 +271,202 @@ func (c *addonDeployController) applyWork(ctx context.Context, required *workapi
 	return nil, err
 }
 
-func buildManifestWorkFromObject(cluster, addonName string, objects []runtime.Object) (*workapiv1.ManifestWork, error) {
-	work := &workapiv1.ManifestWork{
+func buildManifestWorkFromObject(
+	cluster, addonName string,
+	objects []runtime.Object) (deployManifestWork, hookManifestWork *workapiv1.ManifestWork, err error) {
+	var deployManifests []workapiv1.Manifest
+	var hookManifests []workapiv1.Manifest
+	var manifestConfigs []workapiv1.ManifestConfigOption
+
+	for _, object := range objects {
+		rawObject, err := runtime.Encode(unstructured.UnstructuredJSONScheme, object)
+		if err != nil {
+			return nil, nil, err
+		}
+		unstructuredObj := &unstructured.Unstructured{}
+		err = unstructuredObj.UnmarshalJSON(rawObject)
+		if err != nil {
+			return nil, nil, err
+		}
+
+		if isPreDeleteHookObject(unstructuredObj) {
+			hookManifests = append(hookManifests, workapiv1.Manifest{
+				RawExtension: runtime.RawExtension{Raw: rawObject},
+			})
+			manifestConfigs = append(manifestConfigs, newManifestConfigOption(unstructuredObj))
+		} else {
+			deployManifests = append(deployManifests, workapiv1.Manifest{
+				RawExtension: runtime.RawExtension{Raw: rawObject},
+			})
+		}
+	}
+	deployWorkName := fmt.Sprintf("addon-%s-deploy", addonName)
+	deployManifestWork = newDeployManifestWork(deployWorkName, addonName, cluster, deployManifests)
+	if len(hookManifests) != 0 {
+		hookWorkName := fmt.Sprintf("addon-%s-pre-delete", addonName)
+		hookManifestWork = newHookManifestWork(hookWorkName, addonName, cluster, hookManifests, manifestConfigs)
+		return deployManifestWork, hookManifestWork, nil
+	}
+
+	return deployManifestWork, nil, nil
+}
+
+func ManifestsEqual(new, old []workapiv1.Manifest) bool {
+	if len(new) != len(old) {
+		return false
+	}
+
+	for i := range new {
+		if !equality.Semantic.DeepEqual(new[i].Raw, old[i].Raw) {
+			return false
+		}
+	}
+	return true
+}
+
+func newDeployManifestWork(workName, addonName, clusterName string, manifests []workapiv1.Manifest) *workapiv1.ManifestWork {
+	return &workapiv1.ManifestWork{
 		ObjectMeta: metav1.ObjectMeta{
-			Name:      fmt.Sprintf("addon-%s-deploy", addonName),
-			Namespace: cluster,
+			Name:      workName,
+			Namespace: clusterName,
 			Labels: map[string]string{
 				constants.AddonLabel: addonName,
 			},
 		},
 		Spec: workapiv1.ManifestWorkSpec{
 			Workload: workapiv1.ManifestsTemplate{
-				Manifests: []workapiv1.Manifest{},
+				Manifests: manifests,
 			},
 		},
 	}
+}
 
-	for _, object := range objects {
-		rawObject, err := runtime.Encode(unstructured.UnstructuredJSONScheme, object)
-		if err != nil {
-			return nil, err
+func newManifestConfigOption(obj *unstructured.Unstructured) workapiv1.ManifestConfigOption {
+	var resource string
+	var jsonPath workapiv1.JsonPath
+	gvk := obj.GroupVersionKind()
+	switch gvk.Kind {
+	case "Job":
```

Comment:
I think we should move this as part of WellKnownStatus in work agent

Reply:
done